### ISSUE TITLE
Отписываться от BLE характеристика сразу, при вызове startCancel

### ISSUE THREAD

**Issue by AseevEIDev** [Monnoroch/blerpc-android#14]
Судя по коду увидел такую вещь, что отписка от BLE характеристики начнется только тогда, когда придет значение по подписке и в этот момент не будет активных подписок или произойдет ошибка.

Пример 1: у нас много mAh и на 1% мы можем работать 1 час. Мы подписались на батарею, через 5 минут вызвали startCancel() и забыли. А подписка будет все еще активна целый час.

Пример 2: мы подписались на получение номеров блоков прошивки, после чего посылаем метаданные прошивки. И тут метаданные не подходят и мы отменяем перепрошивку и отписываемся от перепрошивки, что вызовет startCancel(). А номер блока никогда не придет и подписка будет активна до тех пор, пока мы будем подключены.

Не знаю, влияет ли активная подписка, по которой не приходят значения на батарею и стоит ли думать о закрытии этой фичи или это ни на что не влияет

**Comment by Monnoroch** [comment]
Я об этом думал и не придумал. Надо спроектировать, прежде, чем что-то делать. Зависеть от канала контроллер не может. Возможно, стоит сделать некий GC, который на все эвенты отписывает лишние подписки.

**Comment by AseevEIDev** [comment]
Надо с Игорем или Сергеем поговорить, влияет ли это хоть как-то на батарею, ведь значения не приходят. 

С другой стороны вдруг появится подписка, которая хоть и не будет ничего посылать, но будет включать какой-нибудь сенсор или что-нибудь считать в это время.

Проблема в том, что при startCancel(), кроме изменения boolean ничего не происходит и если у нас нет других подписок или вызовов, то и эвентов других тоже не будет.

> Зависеть от канала контроллер не может.

Не может, но может иметь Observable аналогично нашим классам для уведомления в приложении и метод для его получения. BleRpcChannel при SUBSCRIBE подписывается на cancelObservable и при отмене сразу же делает отписку от BLE характеристики проверив прежде, что других подписок нет

**Comment by Monnoroch** [comment]
> но может иметь Observable

Нееее, ненене. Мы не будем добавлять в эту систему еще и потоки RX. Это из суперлегкой библиотеки с нулевым оверхедом превратит ее в монстра.

**Comment by AseevEIDev** [comment]
Ну вместо Observable можно использовать Future или что-нибудь в этом роде, чтобы не подключать библиотеки

**Comment by Monnoroch** [comment]
Это не меняет дела, ты подключаешь сложный движок асинхронного исполнения кода с пулами потоков и прочей мутью. В общем, сделать GC -- это ок, можно даже использовать #8 для этого GC, чтоб было больше событий, но добавлять логику в контроллер -- не.

**Comment by Monnoroch** [comment]
Лучше уж зависимость контроллера от канала, при условии, что мы сделаем API нормальное у этого всего.

**Comment by AseevEIDev** [comment]
> GC -- это ок, можно даже использовать #8 для этого GC

Я правильно понимаю, сделать отдельный запрос GC (аналогично System.gc()), как WRITE, RSSI (в будущем), который будет проверять есть ли отмененные подписки и их отписывать?

Зависимость контроллера от канала - это сложнее, учитывая дизайн приложения, которое вообще доступ к каналу не имеет

**Comment by Monnoroch** [comment]
Нет, не запрос. Просто метод, который на все события будет подчистку делать.

**Comment by AseevEIDev** [comment]
Реализовал метод, который вызывается на разные события и вызывает unsubscribe на пустые подписки. Но как оказалось этого недостаточно.

В данный момент на девайсе пульс присылается только если браслет его смог померить. Таким образом если мы получили пульс, отписались и сняли браслет, то браслет будет мерить пульс, пока не изменится уровень батареи или не будет какого-нибудь еще действия.

Это недопустимо, так как:
- при измерении горит красный диод и он не гаснет
- это тратит батарею

Поэтому нужно реализовать мгновенную отписку от характеристики при вызове `startCancel()`. Таким образом `Controller` должен оповестить `BleRpcChannel` о том, что мы отписались. При этом:
- `Controller` о `BleRpcChannel` ничего знать не должен
- мы не используем `RxJava` и `Future`

Таким образом предлагаю добавить в `BleRpcController` методы `setOnUnsubscribeListener(Runnable runnable)` и `removeOnUnsubscribeListener(Runnable runnable)`.
```
public class BleRpcController implements RpcController {
  private final ArrayList<Runnable> cancelCallbacks = new ArrayList();

  @Override
  synchronized public void startCancel() {
    for (Runnable runnable : cancelCallbacks) {
      runnable.run();
    }
  }

  synchronized public void setOnUnsubscribeListener(Runnable runnable) {
    cancelCallbacks.add(runnable);
  }

  synchronized public void removeOnUnsubscribeListener(Runnable runnable) {
    cancelCallbacks.remove(runnable);
  }
}
```

На счет многопоточности, есть несколько способов:
- просто поставить `synchronized`, так как мы синхронизируем один объект - список коллбэков
- ставить замок на объекте `cancelCallbacks`, так как в будущем могут появится другие объекты для синхронизации
```
public void setOnUnsubscribeListener(Runnable runnable) {
    synchronized(cancelCallbacks) {
      cancelCallbacks.add(runnable);
    }
}
```
- использовать `CopyOnWriteArrayList<Runnable>`

Мне кажется самого простого способа `synchronized` здесь достаточно, что думаешь на счет всего этого?

**Comment by Monnoroch** [comment]
Мы не будем добавлять в контроллер массивы. Послушай, давай мы будем улучшать интерактивно. Почини сначала большие вещи, потом подумаем про более мелкие.

**Comment by AseevEIDev** [comment]
> Почини сначала большие вещи, потом подумаем про более мелкие.

Ну вот мне Стас пишет, что он пульс померил, а красный диод все еще горит, хотя должен потухнуть и это должно быть видно. Я конечно могу соорудить какой-нибудь костыль, но не хочется же. 

Поэтому я и поднимаю эту тему, так как она сейчас очень актуальна.

**Comment by Monnoroch** [comment]
В таком случае должна быть зависимость от канала, иначе никак.

**Comment by Monnoroch** [comment]
То, что ты сделал -- это костыль. Если ты говоришь, что отписка должна быть активной, а не eventually, то иначе никак.

**Comment by AseevEIDev** [comment]
Ок, значит `BleRpcController` должен иметь в зависимости `BleRpcChannel`. Доступ к `BleRpcChannel` есть только у класса `ServiceStubFactory`, значит туда еще нужно добавить метод `createController(BluetoothDevice device)`?

**Comment by Monnoroch** [comment]
А лучше createControllerForStub(...)

**Comment by Monnoroch** [comment]
Ну и не стоит торопиться. Можно же отдавать и контроллер, например. Надо продумать, как лучше сделать.

**Comment by Monnoroch** [comment]
И у меня вот ещё мысль. Оно называется BLE не без причины. С чего ты взял, что открытая подписка вообще что-то тратит? Это просто один бит, поставленный в единицу на девайсе.

**Comment by AseevEIDev** [comment]
С того, что при измерении пульса, даже если нам ничего не приходит:
- у девайса горит красный диод
- девайс постоянно меряет пульс, он просто его не присылает, так как руки рядом нет, значения явно не те и девайс их фильтрует

Поэтому телефону все равно, а вот девайсу нет

**Comment by Monnoroch** [comment]
А с чего ты решил, что это не баг конкретной прошивки? Или что это не плохо сделанное при девайса?

**Comment by Monnoroch** [comment]
@AseevEIDev обсудили оффлайн и решили, что это:
- Выглядит как баг, хоть и было задумано фичей.
- Не дает пользователю форсировать закрытие подписки, что иногда полезно.
- Фикс не сильно усложняет интерфейс библиотеки.
- Фикс не изменит семантики библиотеки.

Потому стоит сделать.

**Comment by AseevEIDev** [comment]
Еще одно предложение. Еще раз все посмотрел, `ServiceStubFactory` создает и возвращает же Stub сервиса, а стаб имеет метод `getChannel`. Таким образом мы из приложения уже имеем доступ к BleRpcChannel и можем вручную при отписке вызвать 2 метода:

```
.doFinally(() -> {
    rpcController.startCancel();
    ((BleRpcChannel) pulseServiceStub.getChannel()).unsubscribe(rpcController);
});
```

Нам нужно только создать public метод в BleRpcChannel `unsubscribe(RpcController rpcController)`, который пробежится по подпискам, найдет у кого есть такой контроллер и начнет для него отписку. Что думаешь, таким образом мы вообще не трогаем контроллер, а добавляем всего лишь один метод в channel?

**Comment by Monnoroch** [comment]
Таким образом мы усложняем API и заставляем делать каст стремный. А с предыдущим предложением все как бы само получается.

**Comment by Monnoroch** [comment]
С другой стороны, с этой новой инфой можно не перегружать фабрику, а сделать BleRpcController.create(stub);
